File: data-structures/binary-tree/BinaryTree.cpp
#include <iostream>
#include <vector>
using namespace std;

class Node {
public:
  int data;
  Node *left;
  Node *right;

  Node(int value) {
    data = value;
    left = right = nullptr;
  }
};

static int index = -1;

Node *buildTree(vector<int> preorder) {
  index++;

  if (preorder[index] == -1) {
    return nullptr;
  }

  Node *root = new Node(preorder[index]);
  root->left = buildTree(preorder);
  root->right = buildTree(preorder);

  return root;
}

void preorder(Node *root) {
  if (root == nullptr) {
    return;
  }

  cout << root->data << " ";
  preorder(root->left);
  preorder(root->right);
}

void inorder(Node *root) {
  if (root == nullptr) {
    return;
  }

  inorder(root->left);
  cout << root->data << " ";
  inorder(root->right);
}

void postorder(Node *root) {
  if (root == nullptr) {
    return;
  }

  postorder(root->left);
  postorder(root->right);
  cout << root->data << " ";
}

File: data-structures/non-linear/binary-tree/Recursive/BinaryTree.go
package main

import "fmt"

type Node struct {
	data  int
	left  *Node
	right *Node
}

type BinaryTree struct {
	idx int
}

func (bt *BinaryTree) BuildTree(nodes []int) *Node {
	bt.idx++
	if nodes[bt.idx] == -1 {
		return nil
	}

	newNode := &Node{data: nodes[bt.idx]}
	newNode.left = bt.BuildTree(nodes)
	newNode.right = bt.BuildTree(nodes)

	return newNode
}

func preorder(root *Node) {
	if root == nil {
		return
	}
	fmt.Print(root.data, " ")
	preorder(root.left)
	preorder(root.right)
}

func inorder(root *Node) {
	if root == nil {
		return
	}
	inorder(root.left)
	fmt.Print(root.data, " ")
	inorder(root.right)
}

func postorder(root *Node) {
	if root == nil {
		return
	}
	postorder(root.left)
	postorder(root.right)
	fmt.Print(root.data, " ")
}

func levelorder(root *Node) {
	if root == nil {
		return
	}

	queue := []*Node{root}

	for len(queue) > 0 {
		level := len(queue)

		for i := 0; i < level; i++ {
			curr := queue[0]
			queue = queue[1:]
			fmt.Print(curr.data, " ")

			if curr.left != nil {
				queue = append(queue, curr.left)
			}
			if curr.right != nil {
				queue = append(queue, curr.right)
			}
		}
		fmt.Println()
	}
}
